### ISSUE TITLE
post FormDialog to user from webapi trigger.

### ISSUE THREAD

**Issue by zxed** [Microsoft/BotBuilder#2096]
I followed the proactive bot workflow and put something together that sends a message to a user when the webapi end point is triggered. I did this by saving the activity that came in when the user first messaged the bot and then later recalling that object and using it to target the user.

I was also able to kick off a FormDialog when a user messaged the bot (from dialog) with a specific statement - for this, I used context.Call

Is it possible to kick off a FormDialog to a user based on the saved activity object I have? Since the dialog context is not there when coming in from a webapi trigger, I'm not sure how I can get the form to the user.

I've tried using various things but nothing shows up on the user side; specifically, I tried 
await Conversation.SendAsync(message, () => new MakeSomeFormDialog());
and
Chain.From(() => FormDialog.FromForm(SomeForm.BuildForm, FormOptions.PromptInStart));
and
the interactive example - Task Interactive<T>(IDialog<T> form) where T : class
however, I think the interactive example, but I think this is attempting to post to bot?

Basically, the work flow I'm trying to build is
1. Trigger API
2. User receives a simple message (i.e. echoBot)
3. User receives a series of questions (i.e. sandwichBot)

**Comment by willportnoy** [comment]
This is an example of proactively interrupting the dialog stack with another dialog based on some external event:

https://github.com/Microsoft/BotBuilder/blob/master/CSharp/Samples/AlarmBot/Models/ExternalEvent.cs#L29

In your case, maybe you want to start a new conversation?

Also, look at these samples: https://github.com/MicrosoftDX/botFramework-proactiveMessages

**Comment by zxed** [comment]
Awesome - I somehow overlooked the botframework samples; prior to this I had downloaded several other samples, but this pretty much got me going - specifically, startNewDialog.

Now I just need to see how I can use the payload that came in from the API to the actual FormDialog - the startNewDialog also gave me a decent example of not having to use FormDialog and just using PromptDialog instead so maybe I'll use that instead.

**Comment by zxed** [comment]
My original q is answered - but I wanted to reopen for one more clarification.

https://github.com/MicrosoftDX/botFramework-proactiveMessages/blob/master/csharp/startNewDialogWithPrompt/ConversationStarter.cs#L34

interrupts the stack and pushes the new dialog (which has prompts) on top.

If I trigger the api again - it will interrupt the stack and push another dialog on top.

Is there a way to cancel out of the prior dialog if its the same type? or just cancel out of it completely?

I added stack.Reset();
```
using (var scope = DialogModule.BeginLifetimeScope(Conversation.Container, message))
            {
                var botData = scope.Resolve<IBotData>();
                await botData.LoadAsync(CancellationToken.None);
                var stack = scope.Resolve<IDialogStack>();
                stack.Reset();
                //interrupt the stack
                var dialog = new SurveyDialog();
```
and I think that is probably whats needed but I am missing something - after I respond to the 2nd dialog and then enter something else, the bot ends up in an error state...

Will keep looking.

**Comment by willportnoy** [comment]
@zxed what is the exception you see after calling IDialogStack.Reset outside of a dialog, as you're doing?  That seems like it should work.

**Comment by zxed** [comment]
I don't know what I'd catch that exception, but I'll have a look at the output window. 

It does seem logical that it would result in an error though - basically on the user side it shows the prompt with choices, then on the second call it shows the 2nd instance of choices.

I'd think the stack looks something like

2. prompt waiting for input
1. Prompt waiting for input
0. Root waiting for input 

Calling reset before the 2nd re-call; or even the first recall, removes everything including the root that was awaiting input. 

So responding to the latest prompt would be responding to the only dialog on the stack correct? The 2nd answer I would give (which the root dialog might have processed) doesn't know where to send it?

I'll try catching the exception and see if I can find out what's processing my response back that results in the bot saying "it's having issues"

**Comment by zxed** [comment]
I put tin a few break points.

so, with
1. api event triggered
2. activity retrieved
3. something sent to this conversation
4. interrupt with dialog (context.reset before context.call)
5. dialog asks for prompt
 -- Do not respond --
6. api event triggered
7. activity retrieved
8. something sent to this conversation
9. interrupt with dialog (context.reset before context.call)
10. dialog asks for prompt

11. respond
12. response handled
13. say something again (a)
14 exception thrown (b)

for (a) - the break point did make it to the root controller

```
case ActivityTypes.Message:

                        await Conversation.SendAsync(activity, () => new BasicProactiveEchoDialog());
```

I put a break point on the this dialog (this is the main dialog) and it did not make it to the context.wait bit

public Task StartAsync(IDialogContext context)
```

        {

            context.Wait(MessageReceivedAsync);

            return Task.CompletedTask;

        }
``` 

The following exception was thrown in the debug 

Microsoft.Rest.HttpOperationException: Unknown attachment content type
   at Microsoft.Bot.Connector.ErrorHandling.<HandleErrorAsync>d__2`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Connector.ConversationsExtensions.<ReplyToActivityAsync>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Internals.AlwaysSendDirect_BotToUser.<Microsoft-Bot-Builder-Dialogs-Internals-IBotToUser-PostAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Internals.MapToChannelData_BotToUser.<PostAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Internals.LogBotToUser.<Microsoft-Bot-Builder-Dialogs-Internals-IBotToUser-PostAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Bot.Builder.Dialogs.Internals.PostUnhandledExceptionToUserTask.<Microsoft-Bot-Builder-Dialogs-Internals-IPostToBot-PostAsync>d__5`1.MoveNext()
Exception thrown: 'System.InvalidOperationException' in mscorlib.dll

**Comment by zxed** [comment]
So i put this in a  try catch
await Conversation.SendAsync(activity, () => new BasicProactiveEchoDialog());

and the error basically means that "the stack is empty" as I was guessing the issue was...

**Comment by willportnoy** [comment]
@zxed yes, you may have had an unbalanced number of IDialogStack.Call/IDialogStack.Forward and IDialogStack.Done, leading to a stack underflow.

**Comment by zxed** [comment]
but isnt .reset clearing out the stack? - I'm sorry, if I do not follow.
Its a stack, u push stuff onto it, calling reset empty it., wouldn't you expect "the stack is empty" in anycase?

if reset takes it back to 0 - should I be pushing the root dialog back on there?

**Comment by willportnoy** [comment]
@zxed here is an example of calling IDialogStack.Reset: https://github.com/Microsoft/BotBuilder/commit/d13553eff121219c8112d6bb5a8069b0ace60dad

You also have to add a Loop() to prevent popping an empty stack.

**Comment by willportnoy** [comment]
I'm going to close this issue as I think Loop will resolve your issue - feel free to open a new issue for further discussion.
